### ISSUE TITLE
PROCESSOR: Additional harvest costs not being copied correctly

### ISSUE THREAD

**Issue by lbross** [USFS-PNW/Fia-Biosum-Manager#95]
When a scenario is copied, the additional harvest costs are not copied. The configurations for the supplemental harvest costs are copied

**Comment by lbross** [comment]
This is fixed and will be in the next BioSum release

**Comment by lbross** [comment]
@jsfried confirmed this is fixed in v5.7.9. Closing.

**Comment by lbross** [comment]
There are still problems with copying additional harvest costs. In addition to the harvest cost items on the screen not copying correctly, there is a problem with the SQL that copies the values from one scenario to another on the additional_harvest_costs table

**Comment by lbross** [comment]
Fixed the following problems:
1. When copying a Processor scenario, scenario-level harvest cost components were not being copied
2. When copying a Processor scenario, clicking on the 'Update All Values Using Previously Generated Data' generated an error message
3. When copying a Processor scenario, entries on the additional harvest costs table were not copied correctly because the source scenario was not included as a condition in the update SQL.
4. When copying additional harvest costs using either the Update All Values Using Previously Generated Data' button or the 'Assign previously entered values list option,  the additional harvest costs table were not copied correctly because the source scenario was not included as a condition in the update SQL.
5. For items 3 and 4, the previous values weren't cleared out before being copied so if the source scenario was null and the target scenario had a value, the target scenario would incorrectly retain its old value.

**Comment by jsfried** [comment]
additional harvest costs are now being copied. There remains a problem with #2 and #4 in that while you can copy previously generated data from another scenario, it generates this error message, even though the copying does in fact take place:

![image](https://user-images.githubusercontent.com/17485472/35781045-87426ae4-0999-11e8-9a51-89fda43e28ee.png)

A couple of visual annoyances remain: 
1) The first additional cost block continues to be displaced (offset), just as the first record on the wood value screen is displaced-- is this a feature of the framework we are using or a bug in how we are implementing it?
![image](https://user-images.githubusercontent.com/17485472/35781068-e231a514-0999-11e8-8272-97c3cdd06820.png)

2) All of these dialogs for selecting a scenario (to copy additional harvest costs from, or copying a scenario to a new scenario) have the select button outside the visual frame and one has to enlarge the window to find it, which may lead to confusion and be disconcerting for new users and is just plain annoying to experienced users-- extra clicking and dragging for no purpose disrupt flow. Oddly, after adjusting the window boundary to make these buttons visible, I can do any amount of adjusting (even shrinking the window so small that the scenarios to pick from disappear and the window contains nothing by white space-- and these buttons will remain visible). This is a recurring theme throughout BioSum and would be a nice thing to get fixed when there is time to work on the less than essential fixes.

![image](https://user-images.githubusercontent.com/17485472/35781120-979b207e-099a-11e8-96de-a0e4d2253e9b.png)

![image](https://user-images.githubusercontent.com/17485472/35781117-8bde685e-099a-11e8-89be-ff87195849c1.png)

**Comment by lbross** [comment]
@jsfried: Can you provide some more information on the SQL error you received? I tested this thoroughly and wasn't able to generate this error. Project name and processor scenario? And what series of actions did you take to generate this message? I:

1. Open a Processor scenario
2. Go to supplemental harvest costs tab
3. Use the Update All Values button in the lower left-hand corner to trigger the process

Unfortunately I do not see the same visual annoyances you do. I suspect this is due to your use of the larger font option. I can spend some time tracking this down after we get 5.8.0 stabilized but it could take a while because we need to get it looking good for both display options.

**Comment by jsfried** [comment]
Just tried this again on my work computer. I took the 5.7.9 Blue10demo project and opened it in the new 5.8.0 (Jan 30) after first copying the Biosum_ref from the program installation directory under C:\Programsx86)... to the USers\..\Roaming\FIABiosum directory. Went to processor. Made a copy of Scenario 1 into Copy 1, Checked tree_species-- all is OK. Went to Rule defjnitions, Supplemental Harvest Costs, amd added a scenario component (pretend) and populated it. Went to Treatment 100 and edited the first two records from 150 to 45. saved and exited the edit window. Went back and opened the edit window again, they were changed as intended. exited edit again. Selected Assign previously entered values and chose scenario 1 to copy the values from. Got this dialog and said Yes
![image](https://user-images.githubusercontent.com/17485472/35831333-1cff43ba-0a7e-11e8-8af7-9fa3d7c230ae.png)
Then got
![image](https://user-images.githubusercontent.com/17485472/35831352-295e3d8c-0a7e-11e8-8ca5-b130777f1f06.png)

But, all values were restored as intended (45s were overwritten by the 150s). 

Selected assign previously entered values again. Then I got an unhandled exception and a quit button to click. I am unable to reproduce this latter error (the unhandled exception) but the SQL error happens every time.

**Comment by lbross** [comment]
Thank you. Can you elaborate when you say 'Went to Treatment 100 and edited the first two records from 150 to 45' ? I was not aware of a dialog where you could specify by treatment other than sorting the edit screen. This is probably what I'm missing.

**Comment by jsfried** [comment]
If I click Go on this screen, an edit window appears showing only the records for Rx 100. I can edit them at will and save the changes with the disk icon. I will not see records for any other Rx in this edit window, or the scenario wide additional cost I am calling Pretend. 
![image](https://user-images.githubusercontent.com/17485472/35832334-cb86bbf4-0a81-11e8-9b04-12a2939bfaa3.png)

![image](https://user-images.githubusercontent.com/17485472/35832358-e385f6ca-0a81-11e8-87a1-4eef1c6dffb7.png)

**Comment by lbross** [comment]
I was able to replicate the error and fix it! This will be in the next release of 8.5.0.

**Comment by carlinstarrs** [comment]
I'm attempting to test this issue, but when I go into the Processor Module/Start Biosum Processor/scenario1 and press "Copy" at the top, it spins for a while and then the "Copy Processor Scenario" window that comes up is blank.  
![image](https://user-images.githubusercontent.com/12533482/36991995-5b680d12-205e-11e8-9a04-155c3751277f.png)

**Comment by carlinstarrs** [comment]
I create a scenario and named it "scenario2", copied over "scenario1", and then used the "Edit All Values" option to change some of the harvest cost values manually as Jeremy described above. I saved that and then switched to "Assign previously entered values" and assigned values from scenario1. The values were correctly overwritten with scenario1 values.

**Comment by lbross** [comment]
Yes. You have to create the scenario you want to copy to before copying a scenario. Seems backwards at times but that is the way both CORE and PROCESSOR work. Glad it is working!

**Comment by lpotts** [comment]
There was a history on why it was programmed the way it was.

The copy scenario process was a result of an analyst already being in a
scenario and wishing to copy values from another scenario into it.  The message from the user was ‘Why can’t I copy the values from another scenario into this scenario I created’.

We should have thought it through when we developed the enhancement to allow for dual functionality -  1) copy a scenario into an existing scenario (current process) and 2)copy a scenario into a brand new scenario.

-Larry
